util/types/mytime.go

```diff
@@ -270,3 +271,106 @@ func calcWeek(t *mysqlTime, wb weekBehaviour) (year int, week int) {
 	week = days/7 + 1
 	return
 }
+
+const (
+	intervalYEAR        = "YEAR"
+	intervalQUARTER     = "QUARTER"
+	intervalMONTH       = "MONTH"
+	intervalWEEK        = "WEEK"
+	intervalDAY         = "DAY"
+	intervalHOUR        = "HOUR"
+	intervalMINUTE      = "MINUTE"
+	intervalSECOND      = "SECOND"
+	intervalMICROSECOND = "MICROSECOND"
+)
+
+func timestampDiff(intervalType string, t1 TimeInternal, t2 TimeInternal) int64 {
+	seconds, microseconds, neg := calcTimeDiff(t2, t1, 1)
+	months := uint(0)
+	if intervalType == intervalYEAR || intervalType == intervalQUARTER ||
+		intervalType == intervalMONTH {
+		var (
+			yearBeg, yearEnd, monthBeg, monthEnd, dayBeg, dayEnd uint
+			secondBeg, secondEnd, microsecondBeg, microsecondEnd uint
+		)
+
+		if neg {
+			yearBeg = uint(t2.Year())
+			yearEnd = uint(t1.Year())
+			monthBeg = uint(t2.Month())
+			monthEnd = uint(t1.Month())
+			dayBeg = uint(t2.Day())
+			dayEnd = uint(t1.Day())
+			secondBeg = uint(t2.Hour()*3600 + t2.Minute()*60 + t2.Second())
+			secondEnd = uint(t1.Hour()*3600 + t1.Minute()*60 + t1.Second())
+			microsecondBeg = uint(t2.Microsecond())
+			microsecondEnd = uint(t1.Microsecond())
+		} else {
+			yearBeg = uint(t1.Year())
+			yearEnd = uint(t2.Year())
+			monthBeg = uint(t1.Month())
+			monthEnd = uint(t2.Month())
+			dayBeg = uint(t1.Day())
+			dayEnd = uint(t2.Day())
+			secondBeg = uint(t1.Hour()*3600 + t1.Minute()*60 + t1.Second())
+			secondEnd = uint(t2.Hour()*3600 + t2.Minute()*60 + t2.Second())
```

Comment:
`hour * 3600 + minute * 60 + second`  is intuitive, `totalSeconds` doesn't make it better.